### PR TITLE
Adjustments for Python 3.10 (#1588)

### PR DESCRIPTION
**Pull request by kxrob**
Gets pywin32 built and tests running on Python 3.10. 
Pythonwin / win32ui also runs as far as I can see.
For test binaries find wheels in the CI test build artefacts.

Notes:
* Py 3.10 creates versions in .pyd filenames. (May have further consequences not seen so far)
* distutils `bdist_wininst` support (EXE installer) is gone from core Python in Py3.10. 
  And `bdist_msi` may go from the core in 3.11 : https://bugs.python.org/issue39586 .
  Maybe at least `bdist_msi `will be continued in the `setuptools` project.
=> Build an MSI Installer via  `C:\Python310\python setup.py bdist_msi`

### CODE REVIEW

**Comment by kxrob** [comment]
> May be a Python bug regarding "et#"   
( test_win32trace PY_SSIZE_T_CLEAN requried in Py3.10 issue,  obscure crash / freeze far after PyArg_ParseTuple() )

Reported it in: https://bugs.python.org/issue43321
Underway in: https://github.com/python/cpython/pull/24656
doc: https://docs.python.org/3.10/c-api/arg.html : PY_SSIZE_T_CLEAN 

This will turn obscure crashes into nice Python exceptions.

Now there are likely about 25  non-PY_SSIZE_T_CLEAN modules with critical "..#" arg patterns to look after for `Py_ssize_t ` usage:

<details>
  <summary>
-------  ff -r "s#|u#|y#|z#|Z#|et#" -g PY_SSIZE_T_CLEAN *.c* *.i *.h* -e *.pyc -cCvm 8
 </summary>

```
-------  ff -r "s#|u#|y#|z#|Z#|et#" -g PY_SSIZE_T_CLEAN *.c* *.i *.h* -e *.pyc -cCvm 8
:regex = 's#|u#|y#|z#|Z#|et#'
:exclregex = 'PY_SSIZE_T_CLEAN' (ignore case)
:exclude_files add: '*.pyc'
:max recursion depth = 8
:PATTERNS = ['*.c*', '*.i', '*.h*']
:FNREGEX = '(.*\\.c.*|.*\\.i|.*\\.h.*)\\Z(?ms)'
./com/win32com/src/univgw.cpp:690:     if (!PyArg_ParseTuple(args, "Os#:WriteMemory", &obPtr, &pSrc, &size))
./com/win32com/src/univgw_dataconv.cpp:15:     if (!PyArg_ParseTuple(args, "s#:L64", &pSrc, &size))
./com/win32com/src/univgw_dataconv.cpp:30:     if (!PyArg_ParseTuple(args, "s#:UL64", &pSrc, &size))
./com/win32com/src/extensions/PyIPropertyStorage.cpp:320:                         return Py_BuildValue("is#",
./com/win32comext/adsi/src/PyDSOPObjects.cpp:16:     if (!PyArg_ParseTuple(args, "s#:PyStringAsDS_SELECTION_LIST", &sz, &cb))
./com/win32comext/internet/src/internet.cpp:84:     return Py_BuildValue("iiz#", pPD->grfFlags, pPD->dwState, pPD->pData, pPD->cbData);
./com/win32comext/internet/src/PyIInternetSecurityManager.cpp:137:     if (!PyArg_ParseTuple(args, "Olz#l:ProcessUrlAction", &obpwszUrl, &dwAction, &context, &cbcontext, &dwFlags))
./com/win32comext/mapi/src/mapi.i:84: 	if (!PyArg_ParseTuple(args, "s#:HexFromBin", &szData, &dataSize))
./com/win32comext/mapi/src/mapi.i:752: 	if (!PyArg_ParseTuple(args, "s#:MAPIUIDFromBinary", &szVal, &szSize))
./com/win32comext/mapi/src/PyIExchangeManageStore.i:93: 								"y#",
./com/win32comext/mapi/src/PyIExchangeManageStore.i:95: 								"s#",
./com/win32comext/mapi/src/PyIExchangeManageStoreEx.i:91: 								"y#",
./com/win32comext/mapi/src/PyIExchangeManageStoreEx.i:93: 								"s#",
./com/win32comext/mapi/src/PyIMAPIAdviseSink.cpp:20:                 "k(y#iiN)",
   ...
./com/win32comext/mapi/src/PyIMAPISession.i:101:     if(!PyArg_ParseTuple(args,"ls#Ol:OpenMsgStore",&ulParm,&entryString,&entryStrLen, &objIID,&flags))
./com/win32comext/mapi/src/PyIMAPIStatus.i:67:     if (!PyArg_ParseTuple(args, "lz#l:FlushQueues",
#~ ./com/win32comext/mapi/src/mapi_headers/extraMAPIGuids.h:7: // https://docs.microsoft.com/en-us/office/client-developer/outlook/mapi/mapi-constants#outlook-named-properties
#~ -- EXCLREGEX after REGEX match on ./com/win32comext/propsys/src/propsys.cpp:14: #define PY_SSIZE_T_CLEAN
#~ -- EXCLREGEX after REGEX match on ./com/win32comext/shell/src/shell.cpp:16: #define PY_SSIZE_T_CLEAN
#~ ./isapi/src/PyExtensionObjects.cpp:425:     if (!PyArg_ParseTuple(args, "s#|l:WriteClient", &buffer, &buffLenIn, &reserved))
#~ ./isapi/src/PyExtensionObjects.cpp:571:     if (!PyArg_ParseTuple(args, "s#s#|i:SendResponseHeaders", &SendHeaderExInfo.pszStatus, &cchStatus,
#~ ./isapi/src/PyExtensionObjects.cpp:571:     if (!PyArg_ParseTuple(args, "s#s#|i:SendResponseHeaders", &SendHeaderExInfo.pszStatus, &cchStatus,
#~ ./isapi/src/PyExtensionObjects.cpp:878:     if (!PyArg_ParseTuple(args, "OOKsiiz#z#i:TransmitFile",
#~ ./isapi/src/PyExtensionObjects.cpp:878:     if (!PyArg_ParseTuple(args, "OOKsiiz#z#i:TransmitFile",
#~ ./isapi/src/PyFilterObjects.cpp:194:     if (!PyArg_ParseTuple(args, "s#|l:WriteClient", &buffer, &buffLen, &reserved))
#~ ./isapi/src/PyFilterObjects.cpp:1044:         pLog->psz##struct_elem = buf;                                        \
#~ -- EXCLREGEX after REGEX match on ./isapi/src/StdAfx.h:35: #define PY_SSIZE_T_CLEAN
./Pythonwin/ddeconv.cpp:107:     if (!PyArg_ParseTuple(args, "O|z#:Poke", &obCmd, &pData, &dwSize))
./Pythonwin/ddemodule.h:58:         if (helper.call_args("(Nz#)", PyWinObject_FromTCHAR(szItem), pData, dwSize)) {
./Pythonwin/win32ctrlRichEdit.cpp:117:     PyObject *args = Py_BuildValue("(s#)", pbBuff, (Py_ssize_t)cb);
./win32/src/PySID.cpp:19:         if (!PyArg_ParseTuple(args, "s#:SID", &buf, &bufSize)) {
#~ ./win32/src/PyWinTypes.h:273: // Buffer functions that can be used in place of 's#' input format or PyString_AsStringAndSize
./win32/src/PyWinTypesmodule.cpp:378:     if (!PyArg_ParseTuple(args, "s#i", &value, &numBytes, &flags))
#~ ./win32/src/PyWinTypesmodule.cpp:968:     if (PyType_Ready(&Py##type_name) == -1 ||                                     \
#~ ./win32/src/PyWinTypesmodule.cpp:969:         PyDict_SetItemString(dict, #type_name, (PyObject *)&Py##type_name) == -1) \
./win32/src/win32apimodule.cpp:6000:     if (!PyArg_ParseTuple(args, "iis#|iO", &vk, &sc, &state, &statesize, &flags, obhlayout))
./win32/src/win32process.i:1690: 	if (!PyArg_ParseTuple(args, "OOs#:WriteProcessMemory", &obhprocess, &obAddress, &buf, &size))
./win32/src/win32process_swig.cpp:1908: 	if (!PyArg_ParseTuple(args, "OOs#:WriteProcessMemory", &obhprocess, &obAddress, &buf, &size))
#~ -- EXCLREGEX after REGEX match on ./win32/src/win32trace.cpp:33: #define PY_SSIZE_T_CLEAN
./win32/src/wincerapi.i:784: 	if (!PyArg_ParseTuple(args, "Os#|O:CeWriteFile",
./win32/src/win32crypt/PyCRYPTPROV.cpp:333:     if (!PyArg_ParseTupleAndKeywords(args, kwargs, "k|z#", keywords,
./win32/src/win32crypt/win32crypt_structs.cpp:227:         dummy_tuple, obcai, "sz#:CRYPT_ALGORITHM_IDENTIFIER", cai_keys,
./win32/src/win32crypt/win32crypt_structs.cpp:283:     PyObject *ret = Py_BuildValue("{s:k,s:u#}", "ValueType", pcnv->dwValueType, "Value", pcnv->Value.pbData,
./win32/src/win32net/win32netmisc.cpp:1547:         args, kw, "|OOOz#i:NET_VALIDATE_PASSWORD_CHANGE_INPUT_ARG", keywords,
28 FILES FOUND.



----------- ff -r PY_SSIZE_T_CLEAN *.c* *.i *.h* -cCm 8
:regex = 'PY_SSIZE_T_CLEAN'
:max recursion depth = 8
./com/win32comext/propsys/src/propsys.cpp:14: #define PY_SSIZE_T_CLEAN
./com/win32comext/shell/src/shell.cpp:16: #define PY_SSIZE_T_CLEAN
./isapi/src/StdAfx.h:35: #define PY_SSIZE_T_CLEAN
./Pythonwin/win32dc.cpp:17: #define PY_SSIZE_T_CLEAN  // this module is Py_ssize_t clean!
./win32/src/win32clipboardmodule.cpp:18: #define PY_SSIZE_T_CLEAN  // this should be Py_ssize_t clean!
./win32/src/win32trace.cpp:33: #define PY_SSIZE_T_CLEAN
6 FILES FOUND.

```
  details
</details>

(Which were not seens by tests so far. Tests are all green currently. Maybe with the next Python3.10-alpha some already existing tests would bark ...)

Plan: Treat und put #PY_SSIZE_T_CLEAN's individually until the search yields zero results. 
(Only then optionally move #PY_SSIZE_T_CLEAN to pythontypes.h .  Or better not - in case something hided, and we would better let a Python exception popup instead of risking obscure memory corruption again. )

**Comment by arihant2math** [comment]
@kxrob They merged that pr.

File: Pythonwin/stdafx.h [hunk 1]
```diff
@@ -45,7 +45,9 @@ programmers who forget to use the new names. */
 #include "traceback.h"
 #include "pythonrun.h"
 #include "import.h"  // Python: for dynamicattach routines.
+#if PY_VERSION_HEX < 0x030A0000
```
Comment (mhammond) [inline Pythonwin/stdafx.h:48]: I think you can unconditionally kill this.

**Review by mhammond — CHANGES_REQUESTED** [review]
Thanks heaps for this! There's a few nits which I've suggested, and I'd like to understand a couple of the other changes better.

We could also split the `PY_SSIZE_T_CLEAN` etc stuff out too, because everything there is awesome and should just be done regardless. But I don't mind if you want them in the same PR to avoid the git-splitting headaches :)

**Comment by mhammond** [comment]
> We could also split the `PY_SSIZE_T_CLEAN`

I see you rebased recently, so I figured it's probably easier for me to cherry-pick a couple of your commits for this - I'll try and look again soon.

**Comment by mhammond** [comment]
oops - I mean "easier for both of us" :) And that you should be able to rebase and re-force-push again cleanly.

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -50,6 +50,10 @@ jobs:
     - name: Build wheels
       run: |
         python setup.py bdist_wheel --skip-build
+
```
[the remark this answers is not in the document]
Comment (kxrob) [reply .github/workflows/main.yml:53]: "   Py3.10: Drop bdist_wininst in CI
    
    bdist_wininst not supported anymore in Py3.10+"

File: Pythonwin/Win32uiHostGlue.h [hunk 3]
```diff
@@ -211,11 +211,32 @@ inline BOOL Win32uiHostGlue::DynamicApplicationInit(const TCHAR *cmd, const TCHA
         hModWin32ui = LoadLibrary(fname);
     }
     if (!hModWin32ui) {  // sigh - try and import it
+        PyObject* (*pPyImport_ImportModule)(const char *name) = (PyObject* (*)(const char *name))GetProcAddress(hModCore, "PyImport_ImportModule");
```
[the remark this answers is not in the document]
Comment (kxrob) [reply Pythonwin/Win32uiHostGlue.h:214]: Just to "Py3.10: win32ui glue: get module path via Python
    
    .. from win32ui.__file__ .
    Extension DLL file names now include a version string,
    e.g. "win32ui.cp310-win_amd64.pyd".
"

Added a comment in source.

File: Pythonwin/Win32uiHostGlue.h [hunk 4]
```diff
@@ -211,11 +211,32 @@ inline BOOL Win32uiHostGlue::DynamicApplicationInit(const TCHAR *cmd, const TCHA
...
+        //pPy_XDECREF(win32ui);  // dies with the process
+        if (!pyfn) {
+            wsprintf(err_buf, _T("Failed to import win32ui\n"));
+            goto fail_with_error_dlg;
+        }
+        pPyUnicode_AsWideChar(pyfn, fname, MAX_PATH * 2);
+        //pPy_XDECREF(pyfn);  // dies with the process
+        hModWin32ui = GetModuleHandleW(fname);
+/*
         int(__cdecl * pfnPyRun_SimpleString)(const char *);
```
[the remark this answers is not in the document]
Comment (kxrob) [reply Pythonwin/Win32uiHostGlue.h:234]: Done.

File: Pythonwin/stdafx.h (hunk 1 above)
Comment (kxrob) [reply Pythonwin/stdafx.h:48]: Done / trying.

File: Pythonwin/win32uimodule.cpp [hunk 5]
```diff
@@ -564,7 +564,11 @@ int Python_run_command_with_log(const char *command)
     if (m == NULL)
         return -1;
     d = PyModule_GetDict(m);
+#if PY_VERSION_HEX >= 0x030A0000
+    v = PyRun_String(command, Py_file_input, d, d);
```
[the remark this answers is not in the document]
Comment (kxrob) [reply Pythonwin/win32uimodule.cpp:568]: Stripped / trying.

File: setup.py [hunk 6]
```diff
@@ -1035,7 +1035,13 @@ def build_extension(self, ext):
                 name1 = "%s%d%d%s" % (ext.name, sys.version_info[0], sys.version_info[1], extra)
                 name2 = "%s%s" % (ext.name, extra)
             elif ext.name in ("win32ui",):
-                name1 = name2 = ext.name + extra
+                if sys.version_info >= (3, 10):
+                    ##name1 = ext.name + '.cp%s%s-win_%s' % (sys.version_info[0], sys.version_info[1], platform.machine().lower()) + extra
```
[the remark this answers is not in the document]
Comment (kxrob) [reply setup.py:1039]: Done.

File: win32/src/win2krasmodule.cpp [hunk 7]
```diff
@@ -222,12 +222,19 @@ PYWIN_MODULE_INIT_FUNC(win2kras)
 #endif
     // We insist on win32ras being imported - but the least we
     // can do is attempt the import ourselves!
-    HMODULE hmod = GetModuleHandle(modName);
+    HMODULE hmod = NULL; // GetModuleHandle(modName);
```
[the remark this answers is not in the document]
Comment (kxrob) [reply win32/src/win2krasmodule.cpp:225]: "    Py3.10: win2kras: use version in module filename
    
    Get it from win32ras.__file__"

Added source comment. Cleaned.

File: win32/src/win32crypt/win32crypt_structs.cpp [hunk 8]
```diff
@@ -1,6 +1,6 @@
 #include "win32crypt.h"
 // @doc
-PyObject *dummy_tuple = PyTuple_New(0);
+extern PyObject *dummy_tuple;  // = PyTuple_New(0); // set up in win32cryptmodule init
```
[the remark this answers is not in the document]
Comment (kxrob) [reply win32/src/win32crypt/win32crypt_structs.cpp:3]: "  Py3.10: tuple init before module init caused error  "
at runtime AFAIR.
Already init'ed NULL in win32/src/win32crypt/win32cryptmodule.cpp: "PyObject *dummy_tuple = NULL;"

File: win32/test/handles.py [hunk 9]
```diff
@@ -89,7 +89,14 @@ def testOtherHandle(self):
         # is passed to PyWinLong_AsVoidPtr() if its not a string.
         # passing a handle value of '1' should work - there is something
         # at that ordinal
-        win32api.GetProcAddress(sys.dllhandle, h)
+        if sys.version_info >= (3, 10):
```
[the remark this answers is not in the document]
Comment (kxrob) [reply win32/test/handles.py:92]: .

**Comment by kxrob** [comment]
> PY_SSIZE_T_CLEAN       I see you rebased recently, so I figured it's probably easier for both of us if I just cherry-pick a couple of your commits for this - I'll try and look again soon.

Anyway, its not mutually dependent. No urge independent of Py3.10+ either I think.
The comment / drop-down list above has the hopefully rather complete list of remaining unCLEAN locations.  
( Even release with known or unkown unCLEAN locations would just provoke bug reports / PRs in those rare cases with simple Python exceptions ` SystemError "PY_SSIZE_T_CLEAN macro must be defined for '#' formats"` in Py3.10+.  Not affecting Py3.9-  )

**Comment by mhammond** [comment]
Thanks heaps @kxrob, I took many of these commits but also tweaked a few. I think everything here is done, and again, thanks for doing this work.

**Pull request closed without merging**